aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,1040 @@
+            let property_version = *vector::borrow(&property_versions, i);
+            let token_id = token::create_token_id_raw(gtoken.creator, gtoken.collection, token_name, property_version);
+            // check if this token already voted
+            assert!(!bucket_table::contains(&stats.no_votes, &token_id), error::invalid_argument(ETOKEN_ALREADY_VOTED));
+            assert!(!bucket_table::contains(&stats.yes_votes, &token_id), error::invalid_argument(ETOKEN_ALREADY_VOTED));
+
+            // this account owns the token
+            assert!(token::balance_of(signer::address_of(account), token_id) == 1, error::permission_denied(ENOT_OWN_THE_VOTING_DAO_TOKEN));
+            if (vote) {
+                stats.total_yes = stats.total_yes + 1;
+                bucket_table::add(&mut stats.yes_votes, token_id, voter_addr);
+            } else {
+                stats.total_no = stats.total_no + 1;
+                bucket_table::add(&mut stats.no_votes, token_id, voter_addr);
+            };
+            i = i + 1;
+        };
+
+        nft_dao_events::emit_voting_event(
+            voter_addr,
+            nft_dao,
+            proposal_id,
+            vote,
+            token_names,
+            property_versions,
+        );
+    }
+
+    /// Admin can veto an active proposal
+    public entry fun admin_veto_proposal(admin: &signer, proposal_id: u64, nft_dao: address) acquires DAO, Proposals {
+        // validate if proposal is ready to resolve
+        let dao = borrow_global_mut<DAO>(nft_dao);
+        assert!(dao.admin == signer::address_of(admin), error::permission_denied(EINVALID_ADMIN_ACCOUNT));
+        // assert the proposal voting ended
+        let proposals = borrow_global_mut<Proposals>(nft_dao);
+        let proposal = table::borrow_mut(&mut proposals.proposals, proposal_id);
+        let now = timestamp::now_seconds();
+        assert!(now >= proposal.start_time_sec + dao.voting_duration, error::invalid_argument(EPROPOSAL_ENDED));
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+        proposal.resolution = PROPOSAL_VETOED_BY_ADMIN;
+
+        nft_dao_events::emit_resolve_event(
+            proposal_id,
+            nft_dao,
+            proposal.resolution,
+        )
+    }
+
+    /// Anyone can call the resolve function to resolve a proposal.
+    public entry fun resolve<CoinType>(proposal_id: u64, nft_dao: address) acquires Proposals, DAO, ProposalVotingStatistics {
+        resolve_internal<CoinType>(option::none(), proposal_id, nft_dao);
+    }
+
+    /// DAO admin can directly resolve a proposal
+    public entry fun admin_resolve<CoinType>(admin: &signer, proposal_id: u64, nft_dao: address) acquires DAO, Proposals, ProposalVotingStatistics {
+        let resolver = signer::address_of(admin);
+        resolve_internal<CoinType>(option::some(resolver), proposal_id, nft_dao);
```

Comment:
This also only allows an admin to resolve a proposal if voting has ended and if more yes than no votes. If I understand correctly, this function is not useful as at that point, anyone can call resolve to resolve the proposal themselves, regardless of whether they're an admin or not

Reply:
Same problem here. resolve_internal requires that the voting duration has already ended. This is different from my expectation - if the admin needs to wait until after the voting duration is over, it defeats the purpose of having an admin because they can't act on emergency scenarios (e.g. the protocol is hacked and funds are being lost).

Reply:
update the time check here as well.